allennlp_models/vision/models/vilbert_image_retrieval.py

```diff
@@ -0,0 +1,163 @@
+    """
+
+    def __init__(
+        self,
+        vocab: Vocabulary,
+        text_embeddings: TransformerEmbeddings,
+        image_embeddings: ImageFeatureEmbeddings,
+        encoder: BiModalEncoder,
+        pooled_output_dim: int,
+        fusion_method: str = "mul",
+        dropout: float = 0.1,
+        label_namespace: str = "answers",
+        k: int = 1,
+        *,
+        ignore_text: bool = False,
+        ignore_image: bool = False,
+    ) -> None:
+        super().__init__(
+            vocab,
+            text_embeddings,
+            image_embeddings,
+            encoder,
+            pooled_output_dim,
+            fusion_method,
+            dropout,
+            label_namespace,
+            is_multilabel=False,
+            ignore_text=ignore_text,
+            ignore_image=ignore_image,
+        )
+        self.classifier = torch.nn.Linear(pooled_output_dim, 1)
+
+        self.accuracy = CategoricalAccuracy()
+        self.loss = CrossEntropyLoss()
+        self.fbeta = FBetaMeasure(beta=1.0, average="macro")
+
+        self.k = k
+
+    @overrides
+    def forward(
+        self,  # type: ignore
+        box_features: torch.Tensor,
+        box_coordinates: torch.Tensor,
+        box_mask: torch.Tensor,
+        caption: TextFieldTensors,
+        label: torch.Tensor,
+    ) -> Dict[str, torch.Tensor]:
+        batch_size = box_features.shape[0]
+
+        if self.training:
+            # Shape: (batch_size, num_images, pooled_output_dim)
+            pooled_output = self.backbone(box_features, box_coordinates, box_mask, caption)[
+                "pooled_boxes_and_text"
+            ]
+
+            # Shape: (batch_size, num_images)
+            logits = self.classifier(pooled_output).squeeze(-1)
+            probs = torch.softmax(logits, dim=-1)
+
+            outputs = {"logits": logits, "probs": probs}
+            outputs = self._compute_loss_and_metrics(batch_size, outputs, label)
+
+            return outputs
+
+        else:
+            with torch.no_grad():
+                # Shape: (batch_size, num_images, pooled_output_dim)
+                self.backbone.eval()
+                pooled_output = self.backbone(box_features, box_coordinates, box_mask, caption)[
+                    "pooled_boxes_and_text"
+                ]
+                self.backbone.train()
+
+                # Shape: (batch_size, num_images)
+                scores = self.classifier(pooled_output).squeeze(-1)
+
+                # Shape: (batch_size, k)
+                rel_scores, indices = scores.topk(self.k, dim=-1)
+
+                # Shape: (batch_size)
+                pre_logits = torch.sum((indices == label.reshape(-1, 1)), dim=-1).float()
+
+                # Shape: (batch_size, 2)
+                # 1st column == 1 if we found the image in the top k, 2nd column == 1 if we didn't
+                logits = torch.stack((pre_logits, (pre_logits == 0).float()), dim=1)
+
+                outputs = {"logits": logits}
```

Comment:
That's not what logits are. In this code, the "scores" would be the closest thing to logits? I don't know if that term makes sense here at all.

Reply:
Did you call it that for compatibility with some other code?

Reply:
Oh, I see. They are logits in the training code path, and it makes sense there.

Reply:
The `CategoricalAccuracy` metric already supports top-k. Can you just use that instead of rolling your own?

Reply:
Hm. Will that work if I'm using the accuracy without top-k scoring for the training runs? Or should I create a second accuracy metric based on top-k scoring? And if I should create a second one, can you point me at a good example of how to use it? I'm not sure what its inputs should look like in top-k mode.

Reply:
Yes, you'd have to have two different metrics, because the metric forces you to specify `k` in the constructor.